### ISSUE TITLE
[Question] What's the best practice to use Factory in a multi-modules project?

### ISSUE THREAD

**Issue by DevYeom** [hmlongco/Factory#3]
First, thanks for releasing awesome DI library. 👍
I have one question.
I thought as below, but this way seems to have some disadvantages of always having to specify the type as optional since it cannot give an initial value.
If you have feature modules, what do you think is the best way to inject dependencies into each module?

In a feature module,

```swift
class AFeatureContainer: SharedContainer {
    static let optionalService = Factory<ServiceType?> { nil }
}
```

In a main module,

```swift
AFeatureContainer.optionalService.register { Service() }
```

**Comment by sethroot** [comment]
This does seem to be a significant problem. The project I'm working in has a strict hierarchy defined for the various modules where some are only accessible in the main application module. This is done so that we may define a resolver at the top level that can be passed to each module for concrete registration of the various private types we want access to.

The problem I've encountered emerges when we want to begin resolving more complex types with resources from multiple modules.

Going the simplest route: If I attempted to define the Factory instances within the main application module, I would be required to make the types from the other modules `public` in order to fulfill the container registration. This, in theory, would work and allow us to construct these more complex instances within a register call - but I can't sacrifice the usage of private access modifiers to enable this.

Alternatively, I could choose to create stub Factory instances at the main application level (as optional wrappers around public protocols) and then pass the factory to the target module to provide the concrete instances of the private types. This strategy would be fine if I was only resolving types within the target module but in most cases I am going to need registered private types from multiple modules. Passing only a factory makes this not feasible - I'd need the container with a `public static func` which would produce the side effect of having the types produced by that Factory also need to be `public`

This practice does work in a framework like Swinject due to the fact that the registration closure is passed a resolver argument where I am able to resolve concrete instances of private types from other modules via their public interface.

```
import OtherModule

func registerModule(container: Container) {
  container.register(LocalModuleType.self) { resolver in
    let service = resolver.forceResolve(OtherModuleServiceProtocol.self)
    return LocalModuleType(service: service)
  }
}
```

It seems that there is no clear path to achieving the above example without importing whichever module built out the Container, which will likely blow up any sort of isolation within the module hierarchy.

Finally, this issue appears again for the same reason when attempting to use the `@Injected` property wrapper, where I need to directly reference the `public static func defined` on the container.

```
// Forces importing Container from the main app module, producing circular dependency
@Injected(Container.widgetFactory) private var factory: WidgetFactoryProvider!
```

**Comment by hmlongco** [comment]
So I went into an app with several framework modules (Common, Networking) and added Factory as a package to it, and then did the same with the individual modules.

I then did the following, starting with common...
```swift
// Common+Injections.swift

import Foundation
import Factory

public protocol CommonType {}

extension Container {
    public static var commonType = Factory<CommonType> { Common() }
}

private class Common: CommonType {
    public init() {
        print("Common")
    }
}
```
CommonType is a public protocol, and the factory defines an instance of it based on an internal, private type.

Here's the networking module...
```swift
// Networking+Injections.swift

import Foundation
import Common
import Factory

public protocol NetworkType {}

extension Container {
    public static var networkType = Factory<NetworkType> { Network() }
}

private class Network: NetworkType {
    @Injected(Container.commonType) private var commonType
    public init() {
        print("Network")
    }
}
```
NetworkType is a public protocol, and the factory defines an instance of it also based on an internal, private type. It also uses an instance of CommonType.

Then in the app...
```swift
    class CustomCommonType: CommonType {
        init() {
            print("CustomCommonType")
        }
    }

    func testFactory() {
        Container.commonType.register {
            CustomCommonType()
        }
//        Container.Registrations.reset()
        let network = Container.networkType()
    }
```
Running the app logs...
```
CustomCommonType
Network
```
And uncommenting the reset line gives...
```
Common
Network
```
Effectively, any module exporting a protocol service type publishes a public Factory of that type.

The Networking module shown above imports Common to get a dependency of CommonType.... which it would have to do anyway to even *know* about CommonType.

As near as I can see it all hangs together, even to the point of being able to change behavior from outside of the module with  a registration in the main app itself.

**Comment by hmlongco** [comment]
> This does seem to be a significant problem. The project I'm working in has a strict hierarchy defined for the various modules where some are only accessible in the main application module. This is done so that we may define a resolver at the top level that can be passed to each module for concrete registration of the various private types we want access to.

I haven't fully committed to this yet, but checkout the "unsafe" branch of the project which has a few more options.

Here we define an "unsafe" Factory with a type. Attempt to use it before you've registered a valid factory and things will go boom.
```swift
extension Container {
    static let unsafeService = Factory(unsafe: MyServiceType.self)
}

// elsewhere
Container.unsafeService.register { MyService() }
```
You can also approach it from the @Injected side.
```swift
class MyViewModel {
    @Injected(unsafe: MyServiceType.self) var service
}

// elsewhere
Container.shared.register { MyService() as MyServiceType }
```
Here you provide a type and not a path to a container. If you haven't provided a registration in the shared container prior to construction things will also go badly for you.... ;)

Note that shared is now an instance variable that can be passed around, and has a `register` and a `resolve` function, so the following is also valid.
```swift
class MyViewModel {
    var version1: MyServiceType? = Container.shared.resolve()
    var version2: MyServiceType = Container.shared.resolve()!
    static func register(_ container: SharedContainer) {
        container.register { MyService() as MyServiceType }
    }
}
```
Note that in version 1 resolve() returns an optional. If you don't like optionals you can unwrap it as shown in version 2, but you have to do it explicitly in the code.

Bottom line is with these changes you can effectively operate Factory in "Resolver" mode where you *promise* the code that it will have have what it needs when it needs it. But the various "unsafe" parameter tags make it clear that you're doing just that, and that you're taking responsibility for the consequences.

Once you step outside of Factories and go under the hood you're also losing access to scopes and decorators. You'll have to manage such things on your own.

Thoughts?

**Comment by hmlongco** [comment]
Been working with this a bit more and I'm starting to be convinced that you really shouldn't need much more than the original "optional" example or the later public protocol /private implementation mechanism shown below.

If a module exports a service protocol then it should be able to provide a public definition of the factory needed to provide it. That can be nil and optional as shown in the first reply if you really can't expose a type, or by providing a public protocol type back by a private, internal implementation.

```swift
// Common+Injections.swift

import Foundation
import Factory

public protocol CommonType {}

extension Container {
    public static var commonType = Factory<CommonType> { Common() }
}

private class Common: CommonType {
    public init() {
        print("Common")
    }
}
```
I may also provide the "unsafe" factory initializer.
```swift
extension Container {
    public static var unsafeType = Factory(unsafe: CommonType.self)
}
```
But really, if a module can "see" the type/protocol of a service then the code can be structured such that it can see the factory that provides it. 

And by all rights it should be optional if the best you can do is "I promise that someone, somewhere, will register something before you actually try to use it! Really! I promise!"

I don't think Factory needs a "Resolver" mode after all.

**Comment by DevYeom** [comment]
Now i have a clear idea of how to use Factory in multi-modules. Thank you so much for replying in detail with examples. :)

**Comment by sethroot** [comment]
Thank you for the thoughtful response! I will take a look at the changes in the unsafe branch and let you know if that helps to solve some of the issues I had outlined in my first post.

**Comment by hmlongco** [comment]
@sethroot You're welcome. This library is in the early stages and I want to make sure I get it right and that things are working as expected.

I'm still not sure if I want to provide the *unsafe* factory initializer. The primary rationale behind Factory is avoiding the whole "I promise I've provided a registration" approach that Resolver uses.

If the developer can't do the public/private mechanism shown above and *must* provide some hidden registration, then I think the optional mechanism should be the one preferred.
```swift
protocol FeatureAInterface {
    ...
}
extension Container {
    static var featureA = Factory<FeatureAInterface?> { nil }
}
```
Yes, the end result is optional and a bit of a pain to use. But at least it's guaranteed to be safe.

That said, some sneaky developer could still explicitly unwrap the factory...
```swift
let featureA: FeatureAInterface = Container.featureA()!

class WithWrapper {
    @Injected(Container.featureA) var featureA: FeatureAInterface!
}
```
But at least then *they're* the ones that would be doing it, and it wouldn't be something encouraged by Factory.
